### DISCUSSION TITLE
How to measure text size?

### DISCUSSION THREAD

**Discussion by JanDiederich** [pop-os/cosmic-text#163] [Q&A]
Since I couldn't find one, I wrote a util function to measure text size:
```rust
pub fn measure_text(text: &str,
                    attrs: Attrs, buffer: &mut Buffer) -> Option<Size> {
    buffer.set_text(&mut font_system(), text, attrs, Shaping::Advanced);
    buffer.shape_until_scroll(&mut font_system());

    // Get the layout runs
    let layout_runs: LayoutRunIter = buffer.layout_runs();
    let mut run_width: f32 = 0.;
    let mut run_height_low: f32 = f32::max_value();
    let mut run_height_high: f32 = f32::min_value();
    let mut w: u32 = 0;
    println!("======== {} ===========", text);
    for run in layout_runs {
        // Take the max. width of all lines.
        run_width = run_width.max(run.line_w);
        // Take the total height of all lines together.
        run_height_high = run_height_high.max(run.line_y);
        run_height_low = run_height_low.min(run.line_top);

        for g in run.glyphs {
            let mut tmp = swash_cache();
            let img = tmp.get_image(&mut font_system(), g.physical((0., 0.), 1.).cache_key);
            w += &img.clone().unwrap().placement.width;
        }
        // This thing with w works, text has no line-breaks, wrappings.
    }
    println!("w: {}, run_width: {}", w, run_width);
    Size::from_wh(w as f32, run_height_high - run_height_low)
}
```

But this doesn't work, neither `w` nor `run_width` are big enough, at the end of the function. They get bigger as the text gets longer, but they are still smaller than the actual text. So I can't center the text.

What do I do? How do I measure the _actual_ text size?

**Comment by erlendp** [comment]
I think you're probably just missing the glyph offsets. Take a look at the swash draw code at https://github.com/pop-os/cosmic-text/blob/e16b39f29c84773a05457fe39577a602de27855c/src/buffer.rs#L1289

Specifically, you probably want something like this (untested):

```
w = u32::max(w, physical_glyph.x + img.placement.width);
h = u32::max(run.line_y as i32 + physical_glyph.y + img.placement.height);
```

**Reply by JanDiederich** [reply]
Thanks for the answer. I already forgot this question due to its age. Your method isn't really better than the value in `run_width`. It's something like 303.986 with run_width and 304 with the glyphs. While the glyphs operation is much slower.

Plus: I checked again with the current version of the library and run_width seems to be accurate now. I will keep an eye on it, if it's everywhere consistent now.

**Comment by bes** [comment]
Surprised that text measuring isn't a standalone method in cosmic-text, since it is a very useful feature! Thanks for providing a function.
